abi/src/specs.rs

```diff
@@ -0,0 +1,597 @@
+    /// The return type of the message.
+    return_type: ReturnTypeSpec<F>,
+    /// The message documentation.
+    docs: Vec<&'static str>,
+}
+
+mod state {
+    pub struct Selector;
+    pub struct Mutates;
+    pub struct Returns;
+}
+pub struct Missing<S>(PhantomData<fn() -> S>);
+
+impl MessageSpec {
+    pub fn new(
+        name: <MetaForm as Form>::String,
+    ) -> MessageSpecBuilder<
+        Missing<state::Selector>,
+        Missing<state::Mutates>,
+        Missing<state::Returns>,
+    > {
+        MessageSpecBuilder {
+            spec: Self {
+                name,
+                selector: 0,
+                mutates: false,
+                args: vec![],
+                return_type: ReturnTypeSpec::none(),
+                docs: vec![],
+            },
+            marker: PhantomData,
+        }
+    }
+}
+
+pub struct MessageSpecBuilder<Selector, Mutates, Returns> {
+    spec: MessageSpec,
+    marker: PhantomData<fn() -> (Selector, Mutates, Returns)>,
+}
+
+impl<M, R> MessageSpecBuilder<Missing<state::Selector>, M, R> {
+    pub fn selector(self, selector: u32) -> MessageSpecBuilder<state::Selector, M, R> {
+        MessageSpecBuilder {
+            spec: MessageSpec {
+                selector,
+                ..self.spec
+            },
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<S, R> MessageSpecBuilder<S, Missing<state::Mutates>, R> {
+    pub fn mutates(self, mutates: bool) -> MessageSpecBuilder<S, state::Mutates, R> {
+        MessageSpecBuilder {
+            spec: MessageSpec {
+                mutates,
+                ..self.spec
+            },
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<M, S> MessageSpecBuilder<S, M, Missing<state::Returns>> {
+    pub fn returns(
+        self,
+        return_type: ReturnTypeSpec,
+    ) -> MessageSpecBuilder<S, M, state::Returns> {
+        MessageSpecBuilder {
+            spec: MessageSpec {
+                return_type,
+                ..self.spec
+            },
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<S, M, R> MessageSpecBuilder<S, M, R> {
+    /// Sets the input arguments of the event specification.
+    pub fn args<A>(self, args: A) -> Self
+    where
+        A: IntoIterator<Item = MessageParamSpec>,
+    {
+        let mut this = self;
+        debug_assert!(this.spec.args.is_empty());
+        this.spec.args = args.into_iter().collect::<Vec<_>>();
+        this
+    }
+
+    /// Sets the input arguments of the event specification.
+    pub fn docs<D>(self, docs: D) -> Self
+    where
+        D: IntoIterator<Item = &'static str>,
+    {
+        let mut this = self;
+        debug_assert!(this.spec.docs.is_empty());
+        this.spec.docs = docs.into_iter().collect::<Vec<_>>();
+        this
+    }
+}
+
+impl MessageSpecBuilder<state::Selector, state::Mutates, state::Returns> {
+    pub fn done(self) -> MessageSpec {
+        self.spec
+    }
+}
+
+impl IntoCompact for MessageSpec {
```

Comment:
Worth implementing a derive for these `IntoCompact` impls to reduce boilerplate?

Reply:
We could do that but then again the types that have to implement `IntoCompact` is just a limited set and the implementation is not that straight forward.